Bug fixes improvements

kuratsak opened the pull request:
## What does this PR do?
Some minor bug fixes and robustness improvements.
Split from the 


## Where should the reviewer start?
Each commit is a self contained change/fix, can be reviewed independently.
The retries commit should be viewed ignoring whitespace, that will reduce the diff by ~90%, The try + except clause just indented the lines it surrounded.


## How should this PR be tested?
You might notice changes in a few scenarios:
1. supply a bad data set and see that you get a meaningful exception when running tpot (what exception a model had trying to .fit the data).
2. run tpot and see argument list printed correctly: arg_name == arg_value VS index == arg_name
3. run tpot and after optimizing a single pipeline ctrl+C 3-4 times to see tpot doesn't give up and still outputs an optimized pipeline.


## Any background context you want to provide?
These fixes are part of a bigger split pull request which will include 2-3 more pull requests for each separate feature.


## What are the relevant issues?

## Screenshots (if appropriate)



## Questions:

- Do the docs need to be updated? No
- Does this PR add new (Python) dependencies? No

rhiever commented:
@weixuanfu2016 and @teaearlgraycold, please take a look sometime this week. Most of these changes look reasonable.

duboviy commented on tpot/base.py:
```diff
@@ -486,50 +486,65 @@ def pareto_eq(ind1, ind2):
-                                warnings.simplefilter('ignore')
-                                self._pareto_front_fitted_pipelines[str(pipeline)].fit(features, classes)
+                    break
+                except:
```
- To catch any exception use the ```except Exception```, instead of ```except BaseException``` or just ```except```.
- Indicate the most specific exception type in ```except``` block.

For example:
```python
# Bad
try:
    mapping[key]
except Exception:
    ...

# Better
try:
    mapping[key]
except KeyError:
    ...
```

rhiever replied:
Please fix this one was well prior to the merge.

kuratsak commented:
The reason  I specifically wanted this except to be as global as possible
is for the retries of saving an optimised pipeline after days of work to
try his best to save it whatever the error. We can limit it to less, I
don't mind.

On Sun, 30 Apr 2017, 22:48 duboviy, <notifications@github.com> wrote:

> *@duboviy* requested changes on this pull request.
> ------------------------------
>
> In tpot/base.py
> <https://github.com/rhiever/tpot/pull/432#discussion_r114082609>:
>
> >
> -                        for pipeline in self._pareto_front.items:
> -                            self._pareto_front_fitted_pipelines[str(pipeline)] = self._toolbox.compile(expr=pipeline)
> -                            with warnings.catch_warnings():
> -                                warnings.simplefilter('ignore')
> -                                self._pareto_front_fitted_pipelines[str(pipeline)].fit(features, classes)
> +                    break
> +                except:
>
>
>    - To catch any exception use the except Exception, instead of except
>    BaseException or just except.
>    - Indicate the most specific exception type in except block.
>
> For example:
>
> # Badtry:
>     mapping[key]except Exception:
>     ...
> # Bettertry:
>     mapping[key]except KeyError:
>     ...
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rhiever/tpot/pull/432#pullrequestreview-35524031>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AMzW7FJv65Tcwm3hA1x2j3GMrKhbOzlxks5r1OWagaJpZM4NMiJi>
> .
>

teaearlgraycold commented on tpot/base.py:
```diff
@@ -486,50 +486,65 @@ def pareto_eq(ind1, ind2):
-                          'passed the data to TPOT correctly.')
-                else:
-                    self._fitted_pipeline = self._toolbox.compile(expr=self._optimized_pipeline)
+            # keep trying 10 times in case weird things happened like multiple CTRL+C or exceptions
```
At this point I'd say this logic should be put into its own private function.

teaearlgraycold commented on tpot/base.py:
```diff
@@ -486,50 +486,65 @@ def pareto_eq(ind1, ind2):
+                    break
+                except:
+                    # raise the exception if it's our last attempt
+                    if attempt == attempts-1:
```
Put spaces around mathematical operators to conform to PEP8.

rhiever replied:
Please fix this one was well prior to the merge.

weixuanfu2016 commented on tpot/base.py:
```diff
@@ -406,11 +406,11 @@ def fit(self, features, classes, sample_weight=None):
                              'Please confirm that the input data is scikit-learn compatible. '
                              'For example, the features must be a 2-D array and target labels '
-                             'must be a 1-D array.')
+                             'must be a 1-D array.\nException was:{}'.format(str(e)))
```
I think str() is not necessary here. Maybe need a space between " and {

rhiever replied:
Please fix this one prior to the merge.

rhiever commented:
Looks like this PR has merge conflicts now too.

kuratsak commented:
I'll fix them, are the comments here the only questions?

On Fri, 12 May 2017, 18:13 Randy Olson, <notifications@github.com> wrote:

> Looks like this PR has merge conflicts now too.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rhiever/tpot/pull/432#issuecomment-301103984>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AMzW7KccHwdVlk1Vhvb_j32KOqyZPpdeks5r5HblgaJpZM4NMiJi>
> .
>

rhiever commented:
Yep, I think the few issues that folks raised are all that needs fixing prior to the merge.

kuratsak commented:
Awesome, will do first thing when I'm I'm the office.

On Fri, 12 May 2017, 18:24 Randy Olson, <notifications@github.com> wrote:

> Yep, I think the few issues that folks raised are all that needs fixing
> prior to the merge.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rhiever/tpot/pull/432#issuecomment-301107104>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AMzW7HSmAggXSkmB3vMtvTwb8n8pKKQoks5r5HmTgaJpZM4NMiJi>
> .
>

rhiever commented:
ETA on the updates? I plan to review and start merging again for a probable 0.8 release starting next Monday.

kuratsak commented:
Cool, I was hoping to finish it early next week.

On Fri, 19 May 2017, 23:58 Randy Olson, <notifications@github.com> wrote:

> ETA on the updates? I plan to review and start merging again for a
> probable 0.8 release starting next Monday.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rhiever/tpot/pull/432#issuecomment-302809869>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AMzW7PXAq2H9yhBDp2xVMBrwvuPk5jEEks5r7gKOgaJpZM4NMiJi>
> .
>

kuratsak commented:
How soon is the next release? Will I make it adding my actual features into
it? As soon as this PR is merged I was planning to open a PR on each of
those..

On Sat, 20 May 2017, 00:24 Dan Koretsky, <dankoretsky@gmail.com> wrote:

> Cool, I was hoping to finish it early next week.
>
> On Fri, 19 May 2017, 23:58 Randy Olson, <notifications@github.com> wrote:
>
>> ETA on the updates? I plan to review and start merging again for a
>> probable 0.8 release starting next Monday.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/rhiever/tpot/pull/432#issuecomment-302809869>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/AMzW7PXAq2H9yhBDp2xVMBrwvuPk5jEEks5r7gKOgaJpZM4NMiJi>
>> .
>>
>

rhiever commented:
I'm hoping for a release early this upcoming week. Life may get in the way and change that :-) but that's the plan right now.

kuratsak commented:
Cool, I'll try to PR whatever I can..

On Sat, 20 May 2017, 13:56 Randy Olson, <notifications@github.com> wrote:

> I'm hoping for a release early this upcoming week. Life may get in the way
> and change that :-) but that's the plan right now.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rhiever/tpot/pull/432#issuecomment-302865985>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AMzW7L5ZAQ7kDIO-uhL5pYaos6sHkBErks5r7sbJgaJpZM4NMiJi>
> .
>

kuratsak commented:
Fixed all requested comments.
Fixed some additional bugs.
Fixed unit tests which were broken again by someone else (is that a standard occurence?)...

@rhiever 
Please accept my PR ^_^

rhiever commented:
Merged - thanks @kuratsak!
